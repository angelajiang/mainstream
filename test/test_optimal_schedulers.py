import os
import shutil
import subprocess
import pytest

@pytest.mark.unit
@pytest.mark.parametrize("scheduler_type", ["hifi", "stems"])
def test_optimality(scheduler_type):
    output_dir = os.path.join("test", "tmp", scheduler_type)
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)

    schedules_dir = os.path.join(output_dir, "schedules")
    if not os.path.exists(schedules_dir):
        os.makedirs(schedules_dir)
    other_file = os.path.join(schedules_dir, scheduler_type + ".sim.debug.v0")
    exhaustive_file = os.path.join(schedules_dir, "exhaustive.sim.debug.v0")

    subprocess.check_call("test/exhaustive_vs_x_scheduler.sh " + scheduler_type, shell=True)

    other_f1s = []
    with open(other_file, "r") as f:
        for line in f:
            vals = line.split(',')
            metric = round(float(vals[1]), 2)
            other_f1s.append(metric)

    exhaustive_f1s = []
    with open(exhaustive_file, "r") as f:
        for line in f:
            vals = line.split(',')
            metric = round(float(vals[1]), 2)
            exhaustive_f1s.append(metric)

    assert other_f1s == exhaustive_f1s

    shutil.rmtree(output_dir)


@pytest.mark.unit
@pytest.mark.parametrize("scheduler_type,num_apps_range", [
    ("hifi", 5),
    ("stems", 5),
    ("exhaustive", 3),
    ("stems_cpp", 5),
])
def test_regression(regtest, scheduler_type, num_apps_range):
    output_dir = os.path.join("test", "tmp", scheduler_type)
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)

    schedules_dir = os.path.join(output_dir, "schedules")
    if not os.path.exists(schedules_dir):
        os.makedirs(schedules_dir)
    other_file = os.path.join(schedules_dir, scheduler_type + ".sim.debug.v0")

    subprocess.check_call("test/run_scheduler_with_setups.sh {} {}".format(scheduler_type, num_apps_range), shell=True)

    with open(other_file) as f:
        for line in f:
            line = line.split(",")[:-1]
            regtest.write(",".join(line) + "\n")

    shutil.rmtree(output_dir)
