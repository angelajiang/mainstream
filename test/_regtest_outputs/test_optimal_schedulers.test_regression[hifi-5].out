2,0.133914,45,51,4,3,100.0
2,0.126213,51,45,3,4,100.0
2,0.084159,51,33,3,4,100.0
2,0.260686,51,45,3,4,100.0
2,0.196902,33,51,4,2,100.0
3,0.071134,51,33,33,2,4,3,100.0
3,0.21522,33,33,51,3,4,3,100.0
3,0.185122,33,45,45,4,4,3,100.0
3,0.032653,33,57,39,4,4,3,100.0
3,0.048164,51,33,39,3,4,4,100.0
4,0.168909,57,51,69,69,4,4,4,4,100.0
4,0.118502,57,69,69,51,4,4,4,4,100.0
4,0.120092,51,51,63,63,3,4,3,4,100.0
4,0.073375,45,57,57,51,4,3,4,2,100.0
4,0.292296,51,51,51,45,2,3,2,4,100.0
5,0.200025,75,51,69,75,51,4,4,4,4,2,100.0
5,0.173369,69,51,45,69,69,4,3,2,4,4,100.0
5,0.285933,63,63,63,57,63,4,4,3,4,2,100.0
5,0.183529,57,75,75,63,57,4,3,4,3,4,100.0
5,0.229862,75,75,51,63,57,4,4,2,4,4,100.0
